File: src/main/java/de/storchp/opentracks/osmplugin/utils/MapMode.java
package de.storchp.opentracks.osmplugin.utils;

import de.storchp.opentracks.osmplugin.maps.MovementDirection;

public enum MapMode {

    NORTH() {
        public float getHeading(MovementDirection movementDirection) {
            return 0;
        }
    },
    DIRECTION() {
        public float getHeading(MovementDirection movementDirection) {
            return -1 * movementDirection.getCurrentDegrees();
        }
    };

    public abstract float getHeading(MovementDirection movementDirection);

    public static MapMode valueOf(String name, MapMode defaultValue) {
        try {
            return valueOf(name);
        } catch (IllegalArgumentException ignored) {
            return defaultValue;
        }
    }

}

File: src/main/java/de/storchp/opentracks/osmplugin/utils/MapMode.kt
package de.storchp.opentracks.osmplugin.utils

import de.storchp.opentracks.osmplugin.maps.MovementDirection
import java.lang.IllegalArgumentException

enum class MapMode {
    NORTH {
        override fun getHeading(movementDirection: MovementDirection) = 0f
    },
    DIRECTION {
        override fun getHeading(movementDirection: MovementDirection) =
            -1f * movementDirection.currentDegrees
    };

    abstract fun getHeading(movementDirection: MovementDirection): Float
    
}

fun String.toMapMode(defaultValue: MapMode) =
    try {
        MapMode.valueOf(this)
    } catch (_: IllegalArgumentException) {
        defaultValue
    }

